### PR TITLE
[RLlib] Fix LSTM padding for torch model

### LINKED ISSUE #21392: [RLlib] [Bug] LSTM returns incorrect recurrent states
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

RLlib

### What happened + What you expected to happen

Padding affects the recurrent states that are returned from `LSTMWrapper.Forward`. LSTM should use the recurrent state produced at `seq_lens[i]`, rather than the recurrent state at `seq_lens.max()`.

E.g., we have two batches with `seq_lens = [2, 5]`. `input_dict['obs']` is zero-padded to shape `(2,5,feature)`. Then, the inputs and outputs for the LSTM are:
```
obs[b0,t0], obs[b0,t1], 0, 0, 0                            => LSTM => (b0h0, b0c0), (b0h1, b0c1), (b0h2, b0c2), (b0h3, b0c3), (b0h4, b0c4)
obs[b1,t0], obs[b1,t1], obs[b1,t2], obs[b1,t3], obs[b1,t4] => LSTM => (b1h0, b1c0), (b1h1, b1c1), (b1h2, b1c2), (b1h3, b1c3), (b1h4, b1c4)
```
At the end, we return `state = [(b0h4, b0c4), (b1h4, b1c4)]`. We should actually return `state = [(b0h1, b0c1), (b1h4, b1c4)]`. In the current implementation, the zeroth batch state results from feeding three extra zeros to the LSTM.

The error is located here: https://github.com/ray-project/ray/blob/94af7ccc9268a0ebbbfb6b79157f1f42d5925657/rllib/models/torch/recurrent_net.py#L222-L232

I think we can fix this using `torch.nn.utils.rnn.pack_padded_sequence`. I'm starting at Ray later this month, so if you can wait a few weeks I will fix this myself.

### Versions / Dependencies

Latest pip release

### Reproduction script

N/A

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by smorad**
## Why are these changes needed?
The recurrent states used by the torch LSTM model include padding. This will decrease model accuracy if a sequence is less than `max_seq_len`, as the recurrent state at `seq_len` is not used -- the LSTM will see `max_seq_len - seq_len` inputs of zero-padding. See the related issue for an in-depth description.

This should fix this issue, and should potentially increase model accuracy when BPTT spans multiple batches (where the recurrent states are added to the `SampleBatch` to be used in the next training iteration).

## Related issue number
Closes https://github.com/ray-project/ray/issues/21392

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

I ran `for i in {1..2}; do python rllib/examples/cartpole_lstm.py --framework torch --seq-len 19; done` without this fix (red and light blue lines) and with this fix (orange and dark blue) and obtained the following results:

Screen Shot 2022-02-01 at 8 37 31 PM

### CODE REVIEW

**Review by avnishn — COMMENTED** [review]
This looks good, but we're going to need to add your manually run test to the test suite. @sven1977 or I can walk you through that.

**Comment by smorad** [comment]
@avnishn Sure thing! I think I'm finished working for the evening, but please let me know how to do it and I'll give it a go tomorrow.

File: rllib/examples/cartpole_lstm.py [hunk 1]
```diff
@@ -33,6 +33,9 @@
 parser.add_argument(
     "--stop-reward", type=float, default=150.0, help="Reward at which we stop training."
 )
+parser.add_argument(
+    "--seq-len", type=int, default=20, help="Maximum length sequence for BPTT"
```
Comment (gjoliver) [inline rllib/examples/cartpole_lstm.py:37]: --max-seq-len? since it's used that way.

**Review by gjoliver — APPROVED** [review]

**Comment by sven1977** [comment]
Hey @smorad , looks great! There is a small number (2-3) of LSTM related test cases failing. Especially the PPO cartpole learning test. Could you take a look? Let me know, if you need help :)
